src/us/kbase/workspace/database/mongo/MongoWorkspaceDB.java

```diff
@@ -1805,22 +1809,76 @@ private long incrementWorkspaceCounter(final ResolvedWorkspaceID wsidmongo,
 		return newid;
 	}
 	
+	// has a side effect of setting the provid field on each package
 	private void saveProvenance(final List<ObjectSavePackage> packages)
 			throws WorkspaceCommunicationException {
-		final List<MongoProvenance> prov = new LinkedList<MongoProvenance>();
+		final Map<ObjectSavePackage, DBObject> provmap = new HashMap<>();
 		for (final ObjectSavePackage p: packages) {
-			final MongoProvenance mp = new MongoProvenance(
-					p.wo.getProvenance());
-			prov.add(mp);
-			p.mprov = mp;
+			provmap.put(p, new BasicDBObject(toDocument(p.wo.getProvenance())));
 		}
 		try {
-			wsjongo.getCollection(COL_PROVENANCE).insert((Object[])
-					prov.toArray(new MongoProvenance[prov.size()]));
+			wsmongo.getCollection(COL_PROVENANCE).insert(new LinkedList<>(provmap.values()));
 		} catch (MongoException me) {
 			throw new WorkspaceCommunicationException(
 					"There was a problem communicating with the database", me);
 		}
+		for (final ObjectSavePackage p: packages) {
+			p.provid = (ObjectId) provmap.get(p).get(Fields.MONGO_ID); // ew, side effect
+		}
+	}
+
+	private Map<String, Object> toDocument(final Provenance p) {
```

Comment:
I take it that before it just used the map it had.  Now it explicitly creates the document from scratch.  Is that right?

Reply:
What happened before was that Jongo had a built in Jackson object mapper that would map a POJO to BSON (I think) and hand that off to the Mongo client for processing. Since I'm ditching Jongo for various reasons previously described, I couldn't use that any more (and I actually am not a fan of it since they add a bunch of non-default settings to the mapper). I tried to more or less duplicate the mapping strategy with Jackson but couldn't get it to match after a couple hours, so I gave up and did it manually. The latter strategy is what I use in all my new services anyway since you get exactly what you want and it's explicit / no magic. I don't like magic.

Reply:
Got it.  Thanks.